hostresolver/hostresolver.go

```diff
@@ -0,0 +1,89 @@
+package hostresolver
+
+import (
+	"net"
+	"sort"
+	"strconv"
+	"strings"
+	"time"
+
+	"github.com/containous/traefik/log"
+	"github.com/miekg/dns"
+	"github.com/patrickmn/go-cache"
+)
+
+// HostResolver used for host resolver
+type HostResolver struct {
+	Enabled      bool
+	Cache        *cache.Cache
+	ResolvConfig string
+	ResolvDepth  int
+}
+
+// CNAMEResolv used for storing CNAME result
+type CNAMEResolv struct {
+	TTL    int
+	Record string
+}
+
+// CNAMEFlatten check if CNAME records is exists, flatten if possible
+func (hr *HostResolver) CNAMEFlatten(host string) (string, string) {
+	var result []string
+	result = append(result, host)
+	request := host
+	if hr.Cache == nil {
+		hr.Cache = cache.New(30*time.Minute, 5*time.Minute)
+	}
+	rst, found := hr.Cache.Get(host)
+	if found {
+		result = strings.Split(rst.(string), ",")
+	} else {
+		var cacheDuration = 0 * time.Second
+		for i := 0; i < hr.ResolvDepth; i++ {
+			r := hr.CNAMEResolve(request)
+			if r != nil {
+				result = append(result, r.Record)
+				if i == 0 {
+					cacheDuration = time.Duration(r.TTL) * time.Second
+				}
+				request = r.Record
+			} else {
+				break
+			}
+		}
+		hr.Cache.Add(host, strings.Join(result, ","), cacheDuration)
+	}
+	return result[0], result[len(result)-1]
+}
+
+// CNAMEResolve resolve CNAME if exists, and return with the highest TTL
```

Comment:
Can you correct the comment please: CNAMEResolve resolve**s** CNAME if exists, and return**s** with the highest TTL ?